### PR TITLE
update datasource_flow_json_parsers

### PR DESCRIPTION
**Pull request by sanxun0325**
### Describe what this PR does / why we need it
update datasource_flow_json_parsers

### Does this pull request fix one issue?
fix：https://github.com/alibaba/sentinel-golang/issues/191

### Describe how to verify it
go test -v -race ./...

### Special notes for reviews

### CODE REVIEW

**Comment by louyuting** [comment]
Could you please update json array parser for other rules?

**Comment by sanxun0325** [comment]
@louyuting There are still some parts that have not been modified. You can first see whether the modified is ok

**Comment by sanxun0325** [comment]
@louyuting Please help with the review.

File: core/hotspot/rule_manager.go [hunk 1]
```diff
@@ -265,6 +265,9 @@ func IsValidRule(rule *Rule) error {
 	if rule.DurationInSec < 0 {
 		return errors.New("invalid duration")
 	}
+	if len(rule.SpecificItems) == 0 {
```
Comment (louyuting) [inline core/hotspot/rule_manager.go:268]: This check is wrong.
rule.SpecificItems is optional setting for hotspot.Rule.

File: core/flow/rule.go [hunk 2]
```diff
@@ -58,24 +58,24 @@ type ClusterRuleConfig struct {
 // FlowRule describes the strategy of flow control.
 type FlowRule struct {
 	// ID represents the unique ID of the rule (optional).
-	ID uint64
+	ID uint64 `json:"id"`
```
Comment (louyuting) [inline core/flow/rule.go:61]: maybe should add tag: omitempty?

File: core/hotspot/rule.go [hunk 3]
```diff
@@ -85,26 +86,26 @@ func (s *SpecificValue) String() string {
 // Rule represents the hotspot(frequent) parameter flow control rule
 type Rule struct {
 	// Id is the unique id
-	Id string
+	Id string `json:"id"`
```
Comment (louyuting) [inline core/hotspot/rule.go:89]: Same with uppon.
Id is optional

File: core/system/rule.go [hunk 4]
```diff
@@ -57,11 +57,11 @@ func (t AdaptiveStrategy) String() string {
 }
 
 type SystemRule struct {
-	ID uint64
+	ID uint64 `json:"id"`
```
Comment (louyuting) [inline core/system/rule.go:60]: Same with uppon

File: ext/datasource/helper.go [hunk 5]
```diff
@@ -24,32 +25,14 @@ func checkSrcComplianceJson(src []byte) (bool, error) {
 }
 
 // FlowRulesJsonConverter provide JSON  as the default serialization for list of flow.FlowRule
```
Comment (louyuting) [inline ext/datasource/helper.go:27]: Please rename comments also.

File: ext/datasource/helper.go [hunk 6]
```diff
@@ -92,16 +75,8 @@ func SystemRulesJsonConverter(src []byte) (interface{}, error) {
 	}
```
Comment (louyuting) [inline ext/datasource/helper.go:75]: In order to keep naming consistent.
Please rename function name, such as: SystemRuleJsonArrayParser.

File: ext/datasource/helper.go [hunk 7]
```diff
@@ -211,31 +186,9 @@ func HotSpotParamRulesJsonConverter(src []byte) (interface{}, error) {
...
```
Comment (louyuting) [inline ext/datasource/helper.go:186]: In order to keep naming consistent.
Please rename function name, such as:HotSpotParamRuleJsonArrayParser.

File: ext/datasource/helper.go [hunk 8]
```diff
@@ -85,23 +68,15 @@ func NewFlowRulesHandler(converter PropertyConverter) PropertyHandler {
 	return NewDefaultPropertyHandler(converter, FlowRulesUpdater)
 }
 
-// SystemRulesJsonConverter provide JSON  as the default serialization for list of system.SystemRule
-func SystemRulesJsonConverter(src []byte) (interface{}, error) {
+// SystemRulesJsonArrayParser provide JSON  as the default serialization for list of system.SystemRule
+func SystemRulesJsonArrayParser(src []byte) (interface{}, error) {
```
Comment (louyuting) [inline ext/datasource/helper.go:72]: keep naming convention consistent.
Maybe "SystemRuleJsonArrayParser" be better?

File: ext/datasource/helper.go [hunk 9]
```diff
@@ -204,38 +179,16 @@ func NewCircuitBreakerRulesHandler(converter PropertyConverter) *DefaultProperty
 	return NewDefaultPropertyHandler(converter, CircuitBreakerRulesUpdater)
 }
 
-// HotSpotParamRulesJsonConverter decodes list of param flow rules from JSON bytes.
-func HotSpotParamRulesJsonConverter(src []byte) (interface{}, error) {
+// HotSpotParamRulesJsonArrayParser decodes list of param flow rules from JSON bytes.
+func HotSpotParamRulesJsonArrayParser(src []byte) (interface{}, error) {
```
Comment (louyuting) [inline ext/datasource/helper.go:183]: Same thing as up

**Comment by louyuting** [comment]
@sczyh30  could you please take a look at this PR?

**Comment by sczyh30** [comment]
Nice work. Thanks for contributing!

**Pull request merged**